src/nupic/bindings/experimental.i

```diff
@@ -0,0 +1,267 @@
+# define SWIGPY_SLICE_ARG(obj) ((PyObject*) (obj))
+#else
+# define SWIGPY_SLICE_ARG(obj) ((PySliceObject*) (obj))
+#endif
+
+using namespace nupic::experimental::extended_temporal_memory;
+using namespace nupic;
+
+#define CHECKSIZE(var) \
+  NTA_ASSERT(PyArray_DESCR(var)->elsize == 4) << " elsize:" << PyArray_DESCR(var)->elsize
+
+%}
+
+%pythoncode %{
+  uintDType = "uint32"
+%}
+
+%naturalvar;
+
+//--------------------------------------------------------------------------------
+// Extended Temporal Memory
+//--------------------------------------------------------------------------------
+%inline %{
+  template <typename IntType>
+  inline PyObject* vectorToList(const vector<IntType> &cellIdxs)
+  {
+    PyObject *list = PyList_New(cellIdxs.size());
+    for (size_t i = 0; i < cellIdxs.size(); i++)
+    {
+      PyObject *pyIdx = PyInt_FromLong(cellIdxs[i]);
+      PyList_SET_ITEM(list, i, pyIdx);
+    }
+    return list;
+  }
+%}
+
+%extend nupic::experimental::extended_temporal_memory::ExtendedTemporalMemory
+{
+  %pythoncode %{
+    def __init__(self,
+                 columnDimensions=(2048,),
+                 cellsPerColumn=32,
+                 activationThreshold=13,
+                 initialPermanence=0.21,
+                 connectedPermanence=0.50,
+                 minThreshold=10,
+                 maxNewSynapseCount=20,
+                 permanenceIncrement=0.10,
+                 permanenceDecrement=0.10,
+                 predictedSegmentDecrement=0.00,
+                 formInternalConnections=True,
+                 maxSegmentsPerCell=255,
+                 maxSynapsesPerSegment=255,
+                 seed=42):
+      self.this = _EXPERIMENTAL.new_ExtendedTemporalMemory()
+      _EXPERIMENTAL.ExtendedTemporalMemory_initialize(
+        self, columnDimensions, cellsPerColumn, activationThreshold,
+        initialPermanence, connectedPermanence,
+        minThreshold, maxNewSynapseCount, permanenceIncrement,
+        permanenceDecrement, predictedSegmentDecrement, formInternalConnections,
+        seed, maxSegmentsPerCell, maxSynapsesPerSegment)
+
+    def __getstate__(self):
+      # Save the local attributes but override the C++ temporal memory with the
+      # string representation.
+      d = dict(self.__dict__)
+      d["this"] = self.getCState()
+      return d
+
+    def __setstate__(self, state):
+      # Create an empty C++ temporal memory and populate it from the serialized
+      # string.
+      self.this = _EXPERIMENTAL.new_ExtendedTemporalMemory()
+      if isinstance(state, str):
+        self.loadFromString(state)
+        self.valueToCategory = {}
+      else:
+        self.loadFromString(state["this"])
+        # Use the rest of the state to set local Python attributes.
+        del state["this"]
+        self.__dict__.update(state)
+
+    @classmethod
+    def read(cls, proto):
+      instance = cls()
+      instance.convertedRead(proto)
+      return instance
+  %}
+
+  inline PyObject* getActiveCells()
```

Comment:
This copies the TemporalMemory approach in `algorithms.i`. Here's a comment I found explaining why it's this way: https://github.com/numenta/nupic.core/pull/837#discussion_r50487124